dipy/align/tests/test_mattes.py

```diff
@@ -0,0 +1,776 @@
+
+        # Compute the gradient using finite-diferences
+        n = transform.get_number_of_parameters()
+        expected = np.empty_like(actual)
+        for i in range(n):
+            dtheta = theta.copy()
+            dtheta[i] += h
+
+            T = transform.param_to_matrix(dtheta)
+            shape = np.array(static.shape, dtype=np.int32)
+            sp_to_moving = np.linalg.inv(moving_aff).dot(T)
+            samples_moving_grid = (sp_to_moving.dot(samples.T).T)[...,:dim]
+            intensities_moving, inside = interp_method(moving.astype(np.float32),
+                                                       samples_moving_grid)
+            intensities_moving = np.array(intensities_moving, dtype=np.float64)
+            metric.update_pdfs_sparse(intensities_static, intensities_moving)
+            metric.update_mi_metric(update_gradient=False)
+            val1 = metric.metric_val
+            expected[i] = (val1 - val0) / h
+
+        dp = expected.dot(actual)
+        enorm = np.linalg.norm(expected)
+        anorm = np.linalg.norm(actual)
+        nprod = dp / (enorm * anorm)
+        assert_equal(nprod > 0.99, True)
+
+
+def test_sample_domain_regular():
+    # Test 2D sampling
+    shape = np.array((10, 10), dtype=np.int32)
+    affine = np.eye(3)
+    invalid_affine = np.eye(2)
+    sigma = 0
+    dim = len(shape)
+    n = shape[0] * shape[1]
+    k = 2
+    # Verify exception is raised with invalid affine
+    assert_raises(ValueError, sample_domain_regular, k, shape,
+                  invalid_affine, sigma)
+    samples = sample_domain_regular(k, shape, affine, sigma)
+    isamples = np.array(samples, dtype=np.int32)
+    indices = (isamples[:, 0] * shape[1] + isamples[:, 1])
+    # Verify correct number of points sampled
+    assert_array_equal(samples.shape, [n//k, dim])
+    # Verify all sampled points are different
+    assert_equal(len(set(indices)), len(indices))
+    # Verify the sampling was regular at rate k
+    assert_equal((indices%k).sum(), 0)
+
+    # Test 3D sampling
+    shape = np.array((5, 10, 10), dtype=np.int32)
+    affine = np.eye(4)
+    invalid_affine = np.eye(3)
+    sigma = 0
+    dim = len(shape)
+    n = shape[0] * shape[1] * shape[2]
+    k = 10
+    # Verify exception is raised with invalid affine
+    assert_raises(ValueError, sample_domain_regular, k, shape,
+                  invalid_affine, sigma)
+    samples = sample_domain_regular(k, shape, affine, sigma)
+    isamples = np.array(samples, dtype=np.int32)
+    indices = (isamples[:, 0] * shape[1] * shape[2] + \
+               isamples[:, 1] * shape[2] + \
+               isamples[:, 2])
+    # Verify correct number of points sampled
+    assert_array_equal(samples.shape, [n//k, dim])
+    # Verify all sampled points are different
+    assert_equal(len(set(indices)), len(indices))
+    # Verify the sampling was regular at rate k
+    assert_equal((indices%k).sum(), 0)
+
+
+if __name__ == '__main__':
```

Comment:
You don't need to list all the test functions here. You can just say np.testing.run_module_suite() and they will all run from nosetests.